libraries/chain/block_header_state.cpp

```diff
@@ -114,6 +91,121 @@ const proposer_policy& block_header_state::get_last_proposed_proposer_policy() c
    return *it->second;
 }
 
+bool proposed_finalizer_policies_valid(const std::vector<std::pair<block_num_type, finalizer_policy_ptr>>& proposed_fin_policies) {
+   // check that proposed_finalizer_policies is sorted in ascending order without
+   // repeations; this also implies at most *one* proposed finalizer policy exists
+   // for one block number.
+   block_num_type prev_block_num(0);
+
+   for (auto it = proposed_fin_policies.begin(); it != proposed_fin_policies.end(); ++it) {
+      if (it->first < prev_block_num) // must be in strictly ascending order
+         return false;
+      else
+         prev_block_num = it->first;
+   }
+   return true;
+}
+
+// This function evaluates possible promotions from pending to active
+// (removing any pending policies that are known at that time to never become active)
+// and from proposed to pending (removing any proposed policies that are known at that
+// time to never become pending)
+//
+// In particular,
+// 1. If there is a pending policy, determine whether it should be promoted to active.
+//    If the associated block number is less than or equal to the new LIB number,
+//    the pending policy should be promoted to active. This also means the pending slot
+//    is now open for a possible promotion of a proposed policy to pending.
+//    This guarantees that there will be at most one pending policy at any given time.
+// 2. If there is any proposed policy with an associated block number that is less than
+//    or equal to the new LIB number:
+//  i.   Find the proposed policy with the greatest associated block number that is still
+//       less than or equal to the new LIB number (call this the target proposed policy).
+//  ii.  Remove any proposed policies with an associated block number less than that of
+//       the target proposed policy.
+//  iii. If there is no pending policy, promote that target proposed policy to pending.
+//       Otherwise, leave the target proposed policy (and any other proposed policies with
+//       greater associated block numbers) alone in the proposed policy queue.
+//
+void evaluate_finalizer_policies_for_promotion(const block_header_state& prev,
```

Comment:
I was attempting to see if I could come up with a more concise version of this function and here is what I came up with (it passes the tests if you comment out the `generation` update which is new and not supported by the tests):

```
void evaluate_finalizer_policies_for_promotion(const block_header_state& prev,
                                               block_header_state& next_header_state) {
   assert(proposed_finalizer_policies_valid(prev.proposed_finalizer_policies));

   auto lib = next_header_state.core.last_final_block_num();
   auto pending  = prev.pending_finalizer_policy;
   auto proposed = prev.proposed_finalizer_policies;
   auto first_reversible = std::ranges::find_if(proposed, [&](const auto& p) { return p.first > lib; });

   if (pending) {
      if (pending->first <= lib) {
         // promote it to active
         next_header_state.active_finalizer_policy = std::move(pending->second);
         next_header_state.active_finalizer_policy->generation = prev.active_finalizer_policy->generation + 1;
         pending.reset();
      } else {
         if (first_reversible > proposed.begin())
            // if we have more than one irreversible policies in `proposed`, we need to keep only the latest one.
            next_header_state.proposed_finalizer_policies = { first_reversible - 1, proposed.end() };
         else
            next_header_state.proposed_finalizer_policies = std::move(proposed);
         next_header_state.pending_finalizer_policy = std::move(pending);
      }
   }

   if (!pending) {
      // check if a proposed policy is eligible to become pending (i.e. associated with an
      // irreversible block).
      // This will be the policy associated to the largest block number <= lib.
      if (first_reversible > proposed.begin()) {
         next_header_state.pending_finalizer_policy = *(first_reversible - 1);
         next_header_state.pending_finalizer_policy->first = next_header_state.block_num();
         next_header_state.proposed_finalizer_policies = { first_reversible, proposed.end() };
      } else {
         next_header_state.proposed_finalizer_policies = std::move(proposed);
      }
   }
}
```

Reply:
Thanks. I adapted and tried to conform to the algorithm specified in the issue description https://github.com/AntelopeIO/spring/issues/292 as it is easier to follow.

https://github.com/AntelopeIO/spring/pull/307/commits/51e092e293acebd7487a50e34a425cd9a4e37cd3

Reply:
I myself find this version significantly harder to follow.